File: JavaScript/Arrays/Sorted_Array/SortedArray.js
/*******************************************************
 *  SortedArray.js
 *  Created by Stephen Hall on 11/15/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  SortedArray implementation in JavaScript
 *  Update Date      Name            Description
 *  1      10/28/19  Stephen Hall    Updated to class syntax.
 *                                   Tested and fixed bugs
 ********************************************************/

/**
 * SortedArray Class
 */
class SortedArray {
    /**
     * SortedArray Class Constructor
     * @param size: size to initialize the array to
     * @constructor
     */
  constructor(size) {
    if (size !== undefined && size !== null && size > 0)
      this.array = [(this.size = size)];
    else this.array = [(this.size = 4)];
    this.count = 0;
  }

  /**
   * Doubles the size of the internal array
   */
  resize() {
    this.size *= 2;
    var tmp = [size];
    for (var i = 0; i < this.count; i++) {
      tmp[i] = this.array[i];
    }
    this.array = tmp;
  }

  /**
   * Recursive helper method for merge sort
   * @param arr: sub array to be sorted
   * @param l: left index
   * @param r: right index
   */
  mergeSortHelper(arr, l, r) {
    if (l < r) {
      // Same as (l+r)/2, but avoids overflow for
      // large l and h
      var m = l + (r - l) / 2;

      // Sort first and second halves
      this.mergeSortHelper(arr, l, m);
      this.mergeSortHelper(arr, m + 1, r);

      this.merge(arr, l, m, r);
    }
  }

  /**
   * Private helper method for Merger Sort. Merges two sub-arrays of arr[].
   * @param arr: array to be sorted
   * @param l: index of first sub array
   * @param m: merge point
   * @param r: index of second sub array
   */
  merge(left, right) {
    let resultArray = [],
      leftIndex = 0,
      rightIndex = 0;

    // We will concatenate values into the resultArray in order
    while (leftIndex < left.length && rightIndex < right.length) {
      if (left[leftIndex] < right[rightIndex]) {
        resultArray.push(left[leftIndex]);
        leftIndex++; // move left array cursor
      } else {
        resultArray.push(right[rightIndex]);
        rightIndex++; // move right array cursor
      }
    }

    // We need to concat here because there will be one element remaining
    // from either left OR the right
    return resultArray
      .concat(left.slice(leftIndex))
      .concat(right.slice(rightIndex));
  }

  /**
   * Helper method for recursive Quick Sort
   * @param arr: array to be sorted
   * @param low: low index
   * @param high: high index
   */
  quickSortHelper(arr, low, high) {
    if (low < high) {
      // pivot is partitioning index, arr[pivot] is now at right place
      var pivot = this.partition(arr, low, high);

      // Recursively sort elements before and after pivot
      this.quickSortHelper(arr, low, pivot - 1);
      this.quickSortHelper(arr, pivot + 1, high);
    }
  }

  /**
   * Helper method for Quick Sort. Swaps data in the partition
   * @param arr: array to be sorted
   * @param low: low index
   * @param high: high index
   * @return int: pivot index
   */
  partition(arr, low, high) {
    var pivot = arr[high];
    var i = low - 1; // index of smaller element
    for (var j = low; j < high; j++) {
      // If current element is smaller than or
      // equal to pivot
      if (arr[j] <= pivot) {
        i++;
        var temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }
    }
    // swap arr[i+1] and arr[high] (or pivot)
    temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    return i + 1;
  }

  /**
   * Adds new item into the array
   * @param data: Data to add into the array
   * @returns {*}: Data added into the array
   */
  add(data) {
    if (data === null || data === undefined) 
        return null;

    if (this.count === this.size) 
        this.resize();

    this.array[this.count] = data;
    this.count++;

    return data;
  }

  /**
   * Appends the contents of an array to the SortedArray
   * @param data: array to append
   * @returns {boolean}: success|fail
   */
  append(data) {
    if (data !== null && data !== undefined) {
      for (var i = 0; i < data.length; i++) {
        if (data[i] !== null) this.add(data[i]);
      }
      return true;
    }
    return false;
  }

  /**
   * Sets the data at the given index
   * @param index: index to set
   * @param data: data to set index to
   * @returns {boolean}: success|fail
   */
  set(index, data) {
    if (index >= 0 && index < this.size) {
      this.array[index] = data;
      return true;
    }
    return false;
  }

  /**
   * Gets the data at the arrays given index
   * @param index: Index to get data at
   * @returns {*}: Data at the given index or default value of T if index does not exist
   */
  get(index) {
    if (index >= 0 && index < this.size) return this.array[index];
    return null;
  }

  /**
   * Removes the data at arrays given index
   * @param index: Index to remove
   * @returns {*}: Data removed from the array or default T value if index does not exist
   */
  remove(index) {
    if (index < 0 || index > this.count) return null;

    var tmp = this.array[index];
    this.array[index] = null;
    this.count--;
    return tmp;
  }

  /**
   * Resets the internal array to default size with no data
   */
  reset() {
    this.count = 0;
    this.size = 4;
    this.array = [this.size];
  }

  /**
   * Clears all data in the array leaving size intact
   */
  clear() {
    for (var i = 0; i < this.count; i++) {
      this.array[i] = null;
    }
    this.count = 0;
  }

  /**
   * Gets the current count of the array
   * @returns {SortedArray.count|*}: Number of items in the array
   */
  count() {
    return this.count;
  }

  /**
   * Performs Merge Sort on internal array
   * @returns {[*]}: sorted copy of the internal array
   */
  mergeSort(arr) {

    if (arr.length <= 1) {
      return arr;
    }

    const middle = Math.floor(arr.length / 2);
    const left = arr.slice(0, middle);
    const right = arr.slice(middle);

    return this.merge(this.mergeSort(left), this.mergeSort(right));
  }

  /**
   * Performs Bubble sort on internal array
   * @return {[*]}: sorted copy of the internal array
   */
  bubbleSort() {
    var tmp = [this.count];
    for (var i = 0; i < this.count; i++) {
      tmp[i] = this.array[i];
    }

    for (i = 0; i < this.count - 1; i++)
      for (var j = 0; j < this.count - i - 1; j++) {
        if (tmp[j] > tmp[j + 1]) {
          // swap temp and arr[i]
          var temp = tmp[j];
          tmp[j] = tmp[j + 1];
          tmp[j + 1] = temp;
        }
      }
    return tmp;
  }

  /**
   * Performs Quick Sort on the internal array
   * @return {[*]}: sorted copy of the internal array
   */
  quickSort() {
    var tmp = [this.count];
    for (var i = 0; i < this.count; i++) {
      tmp[i] = this.array[i];
    }
    this.quickSortHelper(tmp, 0, this.count - 1);

    return tmp;
  }

  /**
   * Performs Insertion sort on the internal array
   * @return {[*]}: sorted copy of the internal array
   */
  insertionSort() {
    var tmp = [this.count];
    for (var i = 0; i < this.count; i++) {
      tmp[i] = this.array[i];
    }

    for (var i = 0; i < tmp.length; i++) {
      let value = tmp[i];
      for (var j = i - 1; j > -1 && tmp[j] > value; j--) {
        tmp[j + 1] = tmp[j];
      }
      tmp[j + 1] = value;
    }

    return tmp;
  }

  /**
   * Performs Selection Sort on internal array
   * @returns {[*]}: Sorted copy of the internal data
   */
  selectionSort() {
    var tmp = [this.count];
    for (var i = 0; i < this.count; i++) {
      tmp[i] = this.array[i];
    }
    // One by one move boundary of unsorted sub-array
    for (var i = 0; i < this.count - 1; i++) {
      // Find the minimum element in unsorted array
      var min = i;
      for (var j = i + 1; j < this.count; j++) if (tmp[j] < tmp[min]) min = j;

      // Swap the found minimum element with the first element
      var temp = tmp[min];
      tmp[min] = tmp[i];
      tmp[i] = temp;
    }
    return tmp;
  }

  /**
   * prints the array
   */
  print() {
    this.array.map(function(data) {
      if (data !== null) {
        console.log(data);
      }
    });
  }
}

let arr = new SortedArray(10);
arr.append([9, 45, 6, 78, 42, 7, 1]);
arr.print();
console.log("---------bubble------------");
console.log(arr.bubbleSort());
console.log("---------merge-------------");
console.log(arr.mergeSort(arr.array));
console.log("---------insertion---------");
console.log(arr.insertionSort());
console.log("---------quick-------------");
console.log(arr.quickSort());
console.log("---------selection---------");
console.log(arr.selectionSort());

File: Visual_Basic/Arrays/SortedArray/SortedArray.vb
'******************************************************
' *  SortedArray.vb
' *  Created by Stephen Hall on 11/20/17.
' *  Copyright (c) 2017 Stephen Hall. All rights reserved.
' *  SortedArray implementation in Visual Basic
' *******************************************************

Namespace Arrays.SortedArray
	''' <summary>
	''' Sorted Array Class
	''' </summary>
	''' <typeparam name="T">Generic Type</typeparam>
	Public Class SortedArray(Of T As IComparable)
		''' <summary>
		''' Private members
		''' </summary>
		Private _array As T()
		Private _count As Integer
		Private _size As Integer

		''' <summary>
		''' SortedArray default Constructor
		''' </summary>
		Public Sub New()
			_count = 0
			_array = New T(InlineAssignHelper(_size, 4) - 1) {}
		End Sub

		''' <summary>
		''' SortedArray constructor initialized to a specific size
		''' </summary>
		''' <param name="size">Size to initialize the array to</param>
		Public Sub New(size As Integer)
			_count = 0
			_size = If((size > 0), size, 4)
			_array = New T(_size - 1) {}
		End Sub

		''' <summary>
		'''  Doubles the size of the internal array
		''' </summary>
		Private Sub Resize()
			_size *= 2
			Dim tmp As T() = New T(_size - 1) {}
			Array.Copy(_array, tmp, _count)
			_array = tmp
		End Sub

		''' <summary>
		''' Adds new item into the array
		''' </summary>
		''' <param name="data">Data to add into the array</param>
		''' <returns>Data added into the array</returns>
		Public Function Add(data As T) As T
			If data Is Nothing Then
				Return Nothing
			End If

			If _count = _size Then
				Resize()
			End If

			_array(_count) = data
			_count += 1

			Return data
		End Function

		''' <summary>
		''' Appends the contents of an array to the SortedArray
		''' </summary>
		''' <param name="data"> array to append</param>
		''' <returns>success|fail</returns>
		Public Function Append(data As T()) As Boolean
			If data Is Nothing Then
				Return False
			End If
			For Each aData As T In data
				If aData IsNot Nothing Then
					Add(aData)
				End If
			Next
			Return True
		End Function

		''' <summary>
		''' Sets the data at the given index
		''' </summary>
		''' <param name="index">index to set</param>
		''' <param name="data">data to set index to</param>
		''' <returns>success|fail</returns>
		Public Function [Set](index As Integer, data As T) As Boolean
			If index >= 0 AndAlso index < _size Then
				_array(index) = data
				Return True
			End If
			Return False
		End Function

		''' <summary>
		''' Gets the data at the arrays given index
		''' </summary>
		''' <param name="index">Index to get data at</param>
		''' <returns>Data at the given index or default value of T if index does not exist</returns>
		Public Function [Get](index As Integer) As T
			Return If(index >= 0 AndAlso index < _size, _array(index), Nothing)
		End Function

		''' <summary>
		''' Removes the data at arrays given index
		''' </summary>
		''' <param name="index">Index to remove</param>
		''' <returns>Data removed from the array or default T value if index does not exist</returns>
		Public Function Remove(index As Integer) As T
			If index < 0 OrElse index > _count Then
				Return Nothing
			End If

			Dim tmp As T = _array(index)
			_array(index) = Nothing
			_count -= 1
			Return tmp
		End Function

		''' <summary>
		''' Resets the internal array to default size with no data
		''' </summary>
		Public Sub Reset()
			_count = 0
			_size = 4
			_array = New T(_size - 1) {}
		End Sub

		''' <summary>
		''' Clears all data in the array leaving size intact
		''' </summary>
		Public Sub Clear()
			_array = New T(_size - 1) {}
			_count = 0
		End Sub

		''' <summary>
		''' Gets the current count of the array
		''' </summary>
		''' <returns>Number of items in the array</returns>
		Public Function Count() As Integer
			Return _count
		End Function

		''' <summary>
		''' Private helper method for Merger Sort. Merges two sub-arrays of arr[].
		''' </summary>
		''' <param name="arr">array to be sorted</param>
		''' <param name="l1">index of first sub array</param>
		''' <param name="m">merge point</param>
		''' <param name="r2">index of second sub array</param>
		Private Sub Merge(arr As T(), l1 As Integer, m As Integer, r2 As Integer)
			Dim i As Integer, j As Integer, k As Integer
			Dim n1 As Integer = m - l1 + 1
			Dim n2 As Integer = r2 - m

			' create temp arrays
			Dim l3 As T() = New T(n1 - 1) {}
			Dim r4 As T() = New T(n2 - 1) {}

			' Copy data to temp arrays L[] and R[]
			For i = 0 To n1 - 1
				l3(i) = arr(l1 + i)
			Next
			For j = 0 To n2 - 1
				r4(j) = arr(m + 1 + j)
			Next

			' Merge the temp arrays back into arr[l..r]
			i = 0
			' Initial index of first sub-array
			j = 0
			' Initial index of second sub-array
			k = l1
			' Initial index of merged sub-array
			While i < n1 AndAlso j < n2
				If LessThan(l3(i), r4(j)) OrElse EqualTo(l3(i), r4(j)) Then
					arr(k) = l3(i)
					i += 1
				Else
					arr(k) = r4(j)
					j += 1
				End If
				k += 1
			End While

			' Copy the remaining elements of L[], if there are any
			While i < n1
				arr(k) = l3(i)
				i += 1
				k += 1
			End While

			' Copy the remaining elements of R[], if there are any
			While j < n2
				arr(k) = r4(j)
				j += 1
				k += 1
			End While
		End Sub

		''' <summary>
		''' Recursive helper method for merge sort
		''' </summary>
		''' <param name="arr">sub array to be sorted</param>
		''' <param name="l">left index</param>
		''' <param name="r">right index</param>
		Private Sub MergeSortHelper(arr As T(), l As Integer, r As Integer)
			If l < r Then
				' Same as (l+r)/2, but avoids overflow for
				' large l and h
				Dim m As Integer = CType((l + (r - l) / 2), Integer)

				' Sort first and second halves
				MergeSortHelper(arr, l, m)
				MergeSortHelper(arr, (m + 1), r)

				Merge(arr, l, m, r)
			End If
		End Sub

		''' <summary>
		''' Performs Merge Sort on internal array
		''' </summary>
		''' <returns>sorted copy of the internal array</returns>
		Public Function MergeSort() As T()
			Dim tmp As T() = New T(_count - 1) {}
			Array.Copy(_array, tmp, _count)
			MergeSortHelper(tmp, 0, (_count - 1))
			Return tmp
		End Function

		''' <summary>
		''' Performs Bubble sort on internal array
		''' </summary>
		''' <returns>sorted copy of the internal array</returns>
		Public Function BubbleSort() As T()
			Dim tmp As T() = New T(_count - 1) {}
			Array.Copy(_array, tmp, _count)

			For i As Integer = 0 To _count - 2
				For j As Integer = 0 To _count - i - 2
					If GreaterThan(tmp(j), tmp(j + 1)) Then
						' swap temp and arr[i]
						Dim temp As T = tmp(j)
						tmp(j) = tmp(j + 1)
						tmp(j + 1) = temp
					End If
				Next
			Next
			Return tmp
		End Function

		''' <summary>
		''' Helper method for Quick Sort. Swaps data in the partition
		''' </summary>
		''' <param name="arr">array to be sorted</param>
		''' <param name="low">low index</param>
		''' <param name="high">high index</param>
		''' <returns>pivot index</returns>
		Private Function Partition(arr As T(), low As Integer, high As Integer) As Integer
			Dim pivot As T = arr(high)
			Dim temp As T
			Dim i As Integer = (low - 1)
			' index of smaller element
			For j As Integer = low To high - 1
				' If current element is smaller than or
				' equal to pivot
				If LessThan(arr(j), pivot) OrElse EqualTo(arr(j), pivot) Then
					i += 1
					' swap arr[i] and arr[j]
					temp = arr(i)
					arr(i) = arr(j)
					arr(j) = temp
				End If
			Next

			' swap arr[i+1] and arr[high] (or pivot)
			temp = arr(i + 1)
			arr(i + 1) = arr(high)
			arr(high) = temp

			Return i + 1
		End Function

		''' <summary>
		''' Helper method for recursive Quick Sort
		''' </summary>
		''' <param name="arr">array to be sorted</param>
		''' <param name="low">low index</param>
		''' <param name="high">high index</param>
		Private Sub QuickSortHelper(arr As T(), low As Integer, high As Integer)
			If low < high Then
				' pivot is partitioning index, arr[pivot] is now at right place
				Dim pivot As Integer = Partition(arr, low, high)

				' Recursively sort elements before and after pivot
				QuickSortHelper(arr, low, (pivot - 1))
				QuickSortHelper(arr, (pivot + 1), high)
			End If
		End Sub

		''' <summary>
		''' Performs Quick Sort on the internal array
		''' </summary>
		''' <returns>sorted copy of the internal array</returns>
		Public Function QuickSort() As T()
			Dim tmp As T() = New T(_count - 1) {}
			Array.Copy(_array, tmp, _count)
			QuickSortHelper(tmp, 0, (_count - 1))

			Return tmp
		End Function

		''' <summary>
		''' Performs Insertion sort on the internal array
		''' </summary>
		''' <returns>sorted copy of the internal array</returns>
		Public Function InsertionSort() As T()
			Dim tmp As T() = New T(_count - 1) {}
			Array.Copy(_array, 0, tmp, 0, _count)

			For i As Integer = 1 To _count - 1
				Dim key As T = tmp(i)
				Dim j As Integer = i - 1

				' Move elements of arr[0..i-1], that are greater than key, to one position ahead
				While j >= 0 AndAlso LessThan(tmp(j), key)
					tmp(j + 1) = tmp(j)
					j = j - 1
				End While
				tmp(j + 1) = key
			Next
			Return tmp
		End Function

		''' <summary>
		'''  Performs Selection Sort on internal array
		''' </summary>
		''' <returns>Sorted copy of the internal data</returns>
		Public Function SelectionSort() As T()
			Dim tmp As T() = New T(_count - 1) {}
			Array.Copy(_array, tmp, _count)
			' One by one move boundary of unsorted sub-array
			For i As Integer = 0 To _count - 2
				' Find the minimum element in unsorted array
				Dim min As Integer = i
				For j As Integer = i + 1 To _count - 1
					If LessThan(tmp(j), tmp(min)) Then
						min = j
					End If
				Next

				' Swap the found minimum element with the first
				' element
				Dim temp As T = tmp(min)
				tmp(min) = tmp(i)
				tmp(i) = temp
			Next
			Return tmp
		End Function

		''' <summary>
		''' Determines if a is less than b
		''' </summary>
		''' <param name="a">generic type to test</param>
		''' <param name="b">generic type to test</param>
		''' <returns>true|false</returns>
		Private Function LessThan(a As T, b As T) As Boolean
			Return a.CompareTo(b) < 0
		End Function

		''' <summary>
		''' Determines if a is equal to b
		''' </summary>
		''' <param name="a">generic type to test</param>
		''' <param name="b">generic type to test</param>
		''' <returns>true|false</returns>
		Private Function EqualTo(a As T, b As T) As Boolean
			Return a.CompareTo(b) = 0
		End Function

		''' <summary>
		''' Determines if a is greater than b
		''' </summary>
		''' <param name="a">generic type to test</param>
		''' <param name="b">generic type to test</param>
		''' <returns>true|false</returns>
		Private Function GreaterThan(a As T, b As T) As Boolean
			Return a.CompareTo(b) > 0
		End Function
		Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
			target = value
			Return value
		End Function
	End Class
End Namespace